src/components/Setup/Views/ContactInfo.tsx

```diff
@@ -1,21 +1,63 @@
 import { Button, Field } from '@/components/0_Bruddle'
 import { useSetupFlow } from '@/components/Setup/context/SetupFlowContext'
 import { FormProvider, useForm } from 'react-hook-form'
-import { z } from 'zod'
-import { zodResolver } from '@hookform/resolvers/zod'
+import * as yup from 'yup'
+import { yupResolver } from '@hookform/resolvers/yup'
 import { useToast } from '@/components/0_Bruddle/Toast'
+import { useMutation } from '@tanstack/react-query'
+import { useAuth } from '@/context/authContext'
 
-const contactSchema = z.object({
-    email: z.string().email('Please enter a valid email').min(1, 'Email is required'),
-    telegram: z.string().min(1, 'Telegram username is required'),
+const contactSchema = yup.object({
+    contact: yup
+        .string()
+        .required('Please enter an email or Telegram username')
+        .test('isValidContact', 'Please enter a valid email or Telegram username', (value) => {
+            if (!value) return true
+            const isEmail = /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value)
+            const isTelegram = /^@?[\w\d]{5,32}$/.test(value)
+            return isEmail || isTelegram
+        }),
 })
 
-type ContactFormData = z.infer<typeof contactSchema>
+type ContactFormData = yup.InferType<typeof contactSchema>
 
 const ContactInfo = () => {
     const toast = useToast()
+    const { userId } = useAuth()
+
+    const { mutate: updateUser, isPending } = useMutation({
+        mutationKey: ['update-user-', userId],
+        mutationFn: async ({ contact }: ContactFormData) => {
+            const isEmail = /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(contact)
+
+            const response = await fetch('/api/peanut/user/update-user', {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json',
+                },
+                body: JSON.stringify({
+                    email: isEmail ? contact : undefined,
+                    telegram: !isEmail ? contact : undefined,
+                    userId,
+                }),
+            })
+
+            if (!response.ok) {
+                throw new Error('Failed to update contact info')
+            }
+
+            return response.json()
+        },
+        onError: (error) => {
+            console.error('Error updating user', error)
+            toast.error('Failed to update contact info')
+        },
+        onSuccess: () => {
+            handleNext()
+        },
+    })
     const methods = useForm<ContactFormData>({
-        resolver: zodResolver(contactSchema),
+        resolver: yupResolver(contactSchema),
```

Comment:
I've left a comment above ! They're object validation utility libraries that make it easier to validate form. Check them out https://zod.dev/ can also be used in other contexts as well.

PS: I gave up on zod because I couldn't solve the prod vercel issue that came from a obscure compatibility issue 💀 but they do the same job